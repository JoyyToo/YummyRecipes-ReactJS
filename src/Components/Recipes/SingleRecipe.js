import React, {Component} from 'react';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import {Card, CardText,} from 'material-ui/Card';
import FlatButton from 'material-ui/FlatButton';
import {Link} from 'react-router-dom';
import {notify} from 'react-notify-toast'
import axiosInstance from '../Constants/AxiosCall';

class SingleRecipe extends Component {
    // initialize state
    constructor(props) {
        super(props);
        this.state = {
            recipes: [],
            id: "",
            date_created: '',
            date_modified: '',
            name: '',
            time: '',
            ingredients: '',
            procedure: '',
            category_id: '',

        }
    }

    // mount recipes when page loads
    componentDidMount() {
        this.handleRecipe();
    }

    // handle get recipe request
    handleRecipe = (props) => {
        const id = this.props.match.params['id'];
        const category_id = this.props.match.params['category_id'];

        // send GET request to API
        axiosInstance.get(`category/${category_id}/recipes/${id}`)
            .then((response) => {
                let recipe = response.data['recipes'];
                this.setState({
                    date_created: recipe['date_created'],
                    date_modified: recipe['date_modified'],
                    time: recipe['time'],
                    ingredients: recipe['ingredients'],
                    procedure: recipe['procedure'],                                      
                    name: recipe['name'],
                    category_id: recipe['category_id'],
                    id: recipe['id'],
                })
            })

            .catch((error) => {
                notify.show(error.response, 'error', 4000);
            });
    };


    // handle delete recipe request
    handleDeleteRecipe = (event) => {
        let id = event.currentTarget.getAttribute('id');
        const category_id = this.props.match.params['category_id'];

        // send DELETE request to API
        axiosInstance.delete(`category/${category_id}/recipes/${id}`)
            .then((response) => {
                this.handleRecipe();
                notify.show(response.data.message, 'success', 4000);
                this.setState({
                    id: this.state.id
                });
                this.props
                    .history
                    .push(`/categories/${category_id}/recipes`);
            })

            .catch((error) => {
                notify.show(error.response, 'error', 4000);
            });
    };

    // render a single recipe
    render() {
        let recipe = this.state;

        const styles = {
            card2: {
                position: 'relative',
                padding: 10,
                width: 450,
                float: 'left',
                margin: 5,

            },
            align: {
                align: 'left'

            }
        };
        return (
            <div>
                <MuiThemeProvider>

                    <div>
                        {(
                            <div style={styles.card2}>
                                <MuiThemeProvider is="nospace start">
                                    <Card >
                                        <h2 style={{marginLeft: 30, paddingTop: 10}}>{recipe['name']}</h2>
                                        <p style={{marginLeft: 30, marginBottom: 30}}>Time taken: {recipe['time']}</p>
                                        <p style={{marginLeft: 30, marginBottom: 30}}>Ingridients: {recipe['ingredients']}</p>
                                        <p style={{marginLeft: 30, marginBottom: 30}}>Procedure: {recipe['procedure']}</p>
                                        <CardText>
                                            {
                                                <div>

                                                    <p style={{marginLeft: 15}}><i>Created:</i> {this.state.date_created}</p>
                                                    <p style={{marginLeft: 15, marginBottom: 30}}><i>Modified:</i> {this.state.date_modified}</p>

                                                </div>

                                            }
                                        </CardText>
                                     <Link to={"/categories/" + recipe['category_id'] + "/recipes/" + recipe['id']}>
                                                <FlatButton label="Edit"
                                                            id={recipe['id']} desc={recipe['desc']}
                                                            primary={true}/></Link>
                                            <FlatButton label="Delete" style={{'color': 'red'}}
                                                        category_id={recipe['category_id']}
                                                        id={recipe['id']}
                                                        onClick={(event => this.handleDeleteRecipe(event))} />
                                    </Card>

                                </MuiThemeProvider>
                            </div>
                        )}
                    </div>

                </MuiThemeProvider>
            </div>

        )
    }

};

export default SingleRecipe;

