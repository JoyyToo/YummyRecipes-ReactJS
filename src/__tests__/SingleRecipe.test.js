import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import SingleRecipe from '../Components/Recipes/SingleRecipe'

describe('SingleRecipe component', () => {
    const params = {
        match: {
            params: {
                id:1
            }
        }
    }
    const wrapper = shallow(<SingleRecipe match={{params}}/>);
    it('renders properly', () => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
      });
  
      it('renders 1 <SingleRecipe/> component', () =>{
          const component = shallow(<SingleRecipe match={{params}}/>);
          expect(component).toHaveLength(1);
      });
  
      it('it returns props correctly', () =>{
          const component = shallow(<SingleRecipe name="app" match={{params}}/>);
          expect(component.instance().props.name).toBe('app')
      });

    it('contains MuiThemeProvider', () => {
        expect(wrapper.find('MuiThemeProvider')).toHaveLength(2);
    }); 

    it('contains divs', () => {
        expect(wrapper.find('div')).toHaveLength(4);
    });

    it('contains Card', () => {
        expect(wrapper.find('Card')).toHaveLength(1);
    });

    it('contains h2', () => {
        expect(wrapper.find('h2')).toHaveLength(1);
    });

    it('contains p', () => {
        expect(wrapper.find('p')).toHaveLength(5);
    });

    it('contains CardText', () => {
        expect(wrapper.find('CardText')).toHaveLength(1);
    });

    it('contains Link', () => {
        expect(wrapper.find('Link')).toHaveLength(1);
    });

    it('contains FlatButton', () => {
        expect(wrapper.find('FlatButton')).toHaveLength(2);
    });

})