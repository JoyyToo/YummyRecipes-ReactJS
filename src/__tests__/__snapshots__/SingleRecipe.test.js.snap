// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleRecipe component renders properly 1`] = `
<div>
  <MuiThemeProvider>
    <div>
      <div
        style={
          Object {
            "float": "left",
            "margin": 5,
            "padding": 10,
            "position": "relative",
            "width": 450,
          }
        }
      >
        <MuiThemeProvider
          is="nospace start"
        >
          <Card
            expandable={false}
            expanded={null}
            initiallyExpanded={false}
          >
            <h2
              style={
                Object {
                  "marginLeft": 30,
                  "paddingTop": 10,
                }
              }
            />
            <p
              style={
                Object {
                  "marginBottom": 30,
                  "marginLeft": 30,
                }
              }
            >
              Time taken: 
            </p>
            <p
              style={
                Object {
                  "marginBottom": 30,
                  "marginLeft": 30,
                }
              }
            >
              Ingridients: 
            </p>
            <p
              style={
                Object {
                  "marginBottom": 30,
                  "marginLeft": 30,
                }
              }
            >
              Procedure: 
            </p>
            <CardText>
              <div>
                <p
                  style={
                    Object {
                      "marginLeft": 15,
                    }
                  }
                >
                  <i>
                    Created:
                  </i>
                   
                </p>
                <p
                  style={
                    Object {
                      "marginBottom": 30,
                      "marginLeft": 15,
                    }
                  }
                >
                  <i>
                    Modified:
                  </i>
                   
                </p>
              </div>
            </CardText>
            <Link
              replace={false}
              to="/categories//recipes/"
            >
              <FlatButton
                disabled={false}
                fullWidth={false}
                id=""
                label="Edit"
                labelPosition="after"
                labelStyle={Object {}}
                onKeyboardFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onTouchStart={[Function]}
                primary={true}
                secondary={false}
              />
            </Link>
            <FlatButton
              category_id=""
              disabled={false}
              fullWidth={false}
              id=""
              label="Delete"
              labelPosition="after"
              labelStyle={Object {}}
              onClick={[Function]}
              onKeyboardFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onTouchStart={[Function]}
              primary={false}
              secondary={false}
              style={
                Object {
                  "color": "red",
                }
              }
            />
          </Card>
        </MuiThemeProvider>
      </div>
    </div>
  </MuiThemeProvider>
</div>
`;
